docs/source/format/CanonicalExtensions.rst

```diff
@@ -483,6 +483,28 @@ binary values look like.
 
 .. _variant_primitive_type_mapping:
 
+Timestamp With Offset
+=============
+This type represents a timestamp column that stores potentially different timezone offsets per value. The timestamp is stored in UTC alongside the original timezone offset in minutes.
+
+* Extension name: ``arrow.timestamp_with_offset``.
+
+* The storage type of the extension is a ``Struct`` with 2 fields, in order:
+
+  * ``timestamp``: a non-nullable ``Timestamp(time_unit, "UTC")``, where ``time_unit`` is any Arrow ``TimeUnit`` (s, ms, us or ns).
```

Comment:
Why explicitly saying that it should be non-nullable? 

Is that because the nullability can (should) be defined at the struct level, and you want to avoid having an inconsistency between the "timestamp" and "offset_minutes" fields? (e.g. the case where only the "offset_minutes" field would be null for a given row, what does that mean?)

I am only not sure how practical this limitation is in practice. For example when creating a struct from its individual fields, typically the fields itself will contain nulls. Alternatively we could also specify that if one is null, the other should be null as well?

Reply:
> Alternatively we could also specify that if one is null, the other should be null as well?

Yea, that's more or less what I was thinking about. In principle this type only has meaning if both fields are set. To relax these constraints we'd need to come up with a meaning for what a null timestamp and non-null offset would mean and vice versa.

Could be:
- If timestamp is set and offset is null, assume `offset=0`, i.e timestamp is UTC
- If timestamp is null and offset is set, assume the whole value is null (a standalone offset floating around has no meaning)

Or, alternatively:
- If any of the fields is null, assume the whole value is null as well

None of these is inherently better or worse IMHO, it's just a matter of standardizing.

Reply:
I think I prefer the current wording (require nullability to be handled at the struct level) instead of trying to assign semantics to the other combinations.

Reply:
@jorisvandenbossche now I understand what you meant.

One complexity here is that some compute kernels might want to look at just the UTC timestamp field because they only care about the instant, so we should at least warn/recommend what to do when two bitmaps exist. If we make the spec require that the top-level bitmap can be more selective than the inner bitmap, kernels looking at just the timestamp would be allowed to grab the top-level bitmap and apply it to the processing and the output.

@lidavidm I think top-level bitmaps is the best, but inevitably someone will have to make a decision on what to do when more than one bitmap exists and the spec having recommendations could prevent divergence between implementations.

Reply:
One thing we could do with nullable `offset_minutes` field is to have nulls indicate timezone-naive timestamps. As per discussion on the Arrow call - cases with mixed timezone-naive and timezone-aware columns are probably not common, so I'm only bringing this up here for completeness.